docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+
+The user surface initially should look like this:
+
+```console
+instructlab.core.configuration.init
+instructlab.core.taxonomy.diff
+instructlab.core.data.generate
+instructlab.core.data.list
+instructlab.core.model.train
+instructlab.core.model.evaluate
+instructlab.core.model.chat (one shot request probably)
+instructlab.core.model.serve (vLLM and llama-cpp-python)
+instructlab.core.model.download
+instructlab.core.model.list
+instructlab.core.system.info
+instructlab.core.rag.ingest
+instructlab.core.rag.convert
+```
+
+These initial exposed functions can expand to include any new functionality that is more SDK oriented from the various libraries. For example, if SDG adds something like subset selection, teacher as annotator, data mixing, etc we could expose an `instructlab.core.data.annotate` or `instructlab.core.data.mix` that could be invoked in sequence in a user's script with other parts of the ilab workflow. Some things make _less_ sense to be exposed via a CLI, but still are critical to ensuring users get a good model and properly generated data.
+
+There are certain things that only exist in `ilab` currently and functionality that is going to be moving these such as data ingestion, RAG, etc. Forming an SDK for `instructlab` allows us to capture all of these concerns under one API.
+
+These endpoints in combination with the curated InstructLab Config File will open up these workflows to users and allow InstructLab to be easily incorporated into other projects. Allowing people to run things like data generation, and full fine-tuning via an SDK that pulls in their pre-existing config.yaml but also can be run independently will open new avenues for InstructLab adoption and extensibility.
+
+## Changes to the CLI
+
+The `ilab` CLI will need to adapt to this new structure. Commands like `ilab data generate` should, in terms of code, follow this flow:
+
+1. `src/instructlab/cli/data/generate.py`
+2. `src/instructlab/data/generate.py`
+3. `src/instructlab/process.py`
+4. `src/instructlab/api/data/generate.py`
+
+So generally: cli -> internal handling package -> process management package -> api -> library code, is the flow.
+
+The flow of the CLI today is such that the cli package for a command (`src/instructlab/cli/data/generate.py`) parses the command line options, and passes control to the core code (`/src/instructlab/data/generate.py`). This then (if applicable), yields control to the `process` package which kicks off necessary processes for the given command.
+
+The difference with an SDK is that we would eventually want to end up executing `api/data/generate.py`, the actual publicly consumable python SDK. This will ensure that the CLI can do whatever custom handling it needs to do on top, but eventually it must boil down to the `api` package which uses publicly available methods from the various libraries.
```

Comment:
If I understand this right `src/instructlab/data/generate.py` would just kick off processes using the code in `api/data/generate.py`?  Is there a reason that couldn't be done in `cli/data/generate.py`? Is there a reason we need to go from 3 files -> 4 to do the same thing?

Reply:
@cdoern talked about this in-person - basically this decision boils down to whether or not we need process management or not for the SDK